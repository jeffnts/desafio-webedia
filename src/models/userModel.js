import mongoose from 'mongoose'
const Schema = mongoose.Schema
import bcrypt from 'bcryptjs'

const userSchema = new Schema({
    name: {
        type: String,
        required: true
    },
    userName: {
        type: String,
        required: true,
        unique: true
    },
    password:{
        type: String,
        required: true,
        select: false
    }

});

// Auth settings with Brcypt
userSchema.pre('save', function hashPassword(next) {
  if (this.isModified('password')) {
      this.encryptPassword(this.password)
          .then(hash => {
              this.password = hash;
              next();
          })
          .catch(err => next(err));
  } else {
      return next();
  }
});

userSchema.methods = {
  async authenticate(plainText) {
      try {
          return await bcrypt.compare(plainText, this.password);
      } catch (err) {
          return false;
      }
  },
  encryptPassword(password) {
      return bcrypt.hash(password, 8);
  },
};


export default  mongoose.model('User', userSchema)