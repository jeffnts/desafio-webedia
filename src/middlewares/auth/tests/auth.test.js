const app = require('../../../app')
const request = require('supertest')(app)
import userModel from '../../../models/userModel'
const connection = require('../../../config/database/testDB')
import redisClient from '../../../config/server'
const {sanitizeTestObject} = require('../../../config/tests/sanitizeTestObject')

const dropUserCollection = async () => userModel.remove()


let user = {
  '_id': 'TESTOBJECT01',
  'name': 'Jhon',
  'userName': 'jhon',
  'password': '123456'
}

const createUser = async () => userModel.create(user)

beforeEach(() => {
  createUser()
})

afterEach(()=> {
  dropUserCollection()
})

beforeAll(() => {
  dropUserCollection()
})

afterAll(() => {
  dropUserCollection()
  connection.disconect()
})

describe('Auth tests', ()=>{

  test('It should authenticate the user when it is passed the correct userName and password',async () => {

    const res = await request.post('/api/auth/login').send({'userName': user.userName, 'password': user.password})
    expect(sanitizeTestObject(res.body, ['token'])).toMatchSnapshot()
  })

  test('It should not authenticate the user when it is passed the wrong userName ',async () => {

    const res = await request.post('/api/auth/login').send({'userName': 'JÃµao', 'password': user.password})
    expect(sanitizeTestObject(res.body, ['token'])).toMatchSnapshot()
  })

  test('It should not authenticate the user when it is passed the wrong password',async () => {

    const res = await request.post('/api/auth/login').send({'userName': user.userName, 'password': '1'})
    expect(sanitizeTestObject(res.body, ['token'])).toMatchSnapshot()
  })

})