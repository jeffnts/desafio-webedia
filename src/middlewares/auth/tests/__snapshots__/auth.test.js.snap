// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth tests It should authenticate the user when it is passed the correct userName and password 1`] = `
Object {
  "token": "FROZEN-TOKEN",
}
`;

exports[`Auth tests It should not authenticate the user when it is passed the wrong password 1`] = `
Object {
  "message": "Nome de Usuário e Senha são obrigatórios.",
}
`;

exports[`Auth tests It should not authenticate the user when it is passed the wrong userName  1`] = `
Object {
  "message": "Usuário ou senha inválidos!",
}
`;

exports[`Auth tests It should not authenticate the user when it is passed the wrong userName  2`] = `
Object {
  "message": "Usuário ou senha inválidos!",
}
`;
